<h3 style="color: #001a79;">Exercise 1</h3>

<hr style="border-top: 1px solid #001a79;" />

The above gives about a 1.5% chance of randomly selecting the correct cups. Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.

The combinatorial function was gotten from https://stackoverflow.com/a/4941932

In [2]:
#Import math module
import math

In [3]:
math.comb(8,4)

70

In [5]:
math.comb(10,5)

252

In [6]:
#The probability selecting 5 correct cups
1/math.comb(10,5)

0.003968253968253968

If we experiment with 10 cups in total, 5 with milk in first and 5 with tea in first then the chance of selecting randomly and knowing how each was made is 0.3968% which is much less than 1%.

Since the exercise asks for the minimum number of cups required to get a probablility of less than 1% we could try 9 cups in total, 4 with milk in first and 5 with milk in after - or vice versa.

In [8]:
#The number of ways to select 4 cups from 9
math.comb(9,4)

126

In [9]:
#The probability selecting 4 correct cups
1/math.comb(9,4)

0.007936507936507936

Therefore the minimum number of cups required would be 9 in order to have less than 1% chance of picking the correct cups 

*Bonus:* How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?

<h3 style="color: #001a79;">Exercise 2</h3>

<hr style="border-top: 1px solid #001a79;" />

Use <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html" style="color: #ff791e">scipy's version of Fisher's exact test</a> to simulate the Lady Tasting Tea problem.

In [10]:
#import scipy stats module
import scipy.stats as ss

In [11]:
ss.fisher_exact

<function scipy.stats.stats.fisher_exact(table, alternative='two-sided')>

In [15]:
#fishers exact test prints the answer with the oddsratio but I just want the answer so assigned the values 
#variable names and then just asked for the answer back
oddsratio,answer=ss.fisher_exact([[4,0],[0,4]])
answer

0.028571428571428536

Now we need to divide the answer by the number of ways that we can choose the first 4 correct cups, e.g the four cups with milk poured first, because then you will automatically know the next four have milk poured last. So we can divide by '8 choose 4' ways. 

In order to answer this question the following link was used to help gain an understanding https://towardsdatascience.com/fishers-exact-test-from-scratch-with-python-2b907f29e593

In [16]:
answer/math.comb(8,4)

0.0004081632653061219

In [19]:
math.comb(2,1)

2

<h3 style="color: #001a79;">Exercise 3</h3>

Take the code from the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html" style="color: #ff791e">Examples section of the scipy stats documentation for independent samples t-tests</a>, add it to your own notebook and add explain how it works using MarkDown cells and code comments. Improve it in any way you think it could be improved.
<hr style="border-top: 1px solid #001a79;" />

Step 1: Install all necessary packages

In [1]:
#import numpy package
import numpy as np
#import stats package from scipy
from scipy import stats

Step 2:Create a range of random numbers in a wide range of distributions to choose from in the functions to follow

In [3]:
#Create a variable called "rng" which is equal to an array of randomly generated numbers
rng = np.random.default_rng()

Step 3:Create 2 random variable samples, with normal distribution and have identical means and standard deviation. Then run an independent t-test on them.

In [4]:
#Create 2 random variable samples,using the "stats.norm.rvs" function, called "rvs1" and "rvs2"
#The parameters: loc= mean, scale= st dev, size= sample size, random_state= select variables from our rng array above
rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
#Use stats.ttest_ind function to run a t-test on our 2 samples
stats.ttest_ind(rvs1, rvs2)

Ttest_indResult(statistic=0.623566878816196, pvalue=0.5330545541871776)

Results:The t-test statistics is 0.623566878816196 and the corresponding p-value is equal to 0.5330545541871776
<hr style="border-top: 1px solid #001a79;" />
<h4 style="color: #001a79;">Experiment with parameters</h4>

The t-test assumes equal population variance but the t-test often underestimates the p-value for unequal variance

Below the variance is set to unequal but there is no change in results as the mean and st dev of the 2 samples is the same so the variance is equal

In [5]:
#You can set equal_var to False if the population variance is unequal which in this case makes no diff from above
stats.ttest_ind(rvs1, rvs2, equal_var=False)

Ttest_indResult(statistic=0.623566878816196, pvalue=0.5330546694729155)

Compare the results of changing the equal_var parameter using samples with different standard deviations:

In [8]:
#Create a third random variable sample similar to previously except change the loc/mean and scale/st dev
#equal_var=True by default
rvs3 = stats.norm.rvs(loc=5, scale=20, size=500, random_state=rng)
stats.ttest_ind(rvs1, rvs3)

Ttest_indResult(statistic=0.13069772385526884, pvalue=0.8960407725944043)

In [7]:
#Run the test again but change equal_var to equal False
stats.ttest_ind(rvs1, rvs3, equal_var=False)

Ttest_indResult(statistic=0.6336874694151027, pvalue=0.5264818682202548)

Results:This time the p-value changed by increasing when variance was stated as unequal

Next try changing the sample size so they are not equal but have equal means:

In [9]:
#change sample size to 100
rvs4 = stats.norm.rvs(loc=5, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs4)

Ttest_indResult(statistic=1.203274186150214, pvalue=0.22934637575588313)

In [10]:
#run again with equal_var set to false
stats.ttest_ind(rvs1, rvs4, equal_var=False)

Ttest_indResult(statistic=0.8268423459294061, pvalue=0.4100967730135364)

Result:Two random variable samples of different sizes but same means have different t-test and p-value statistics when changing from equal population variance to unequal population variance

Lastly try compare variances of two samples that have different means,st dev and size

In [11]:
#create a new sample with diff mean, scale and size to sample 1
rvs5 = stats.norm.rvs(loc=8, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs5)

Ttest_indResult(statistic=-5.327469436566997, pvalue=1.412311827901223e-07)

In [12]:
#run again with equal_var set to false
stats.ttest_ind(rvs1, rvs5, equal_var=False)

Ttest_indResult(statistic=-3.483338432888841, pvalue=0.0007138234440967376)

Results:As you can see there is an even larger difference between the results for equal population variance vs. unequal population variance
<hr style="border-top: 1px solid #001a79;" />

Note: More permutations mean more accurate results therefore you can insert a permutations parameter

In [13]:
#Insert a permutations parameter equal to 10000
stats.ttest_ind(rvs1, rvs5, permutations=10000,random_state=rng)

Ttest_indResult(statistic=-5.327469436566997, pvalue=0.0)

<hr style="border-top: 1px solid #001a79;" />
<h4 style="color: #001a79;">Example</h4>

In [14]:
#Create 2 samples a and b which are a list of random numbers
a = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)
b = (1.1, 2.9, 4.2)

In [15]:
#run ttest function on the 2 samples
#trim defines the fraction of elements to be trimmed from each end of the samples which is set to 20%
stats.ttest_ind(a, b, trim=.2)

Ttest_indResult(statistic=3.4463884028073513, pvalue=0.01369338726499547)